![Final Lesson Exercise](images/Banner_FEX.png)

# Lesson #2: Data Science Introduction 
## Good chocolate (cont)

## About this assignment
In this assignment, you will continue to explore information regarding good chocolate.<br/>

This time you will do so, using new capabilities, namely web APIs.

## Preceding Step - import modules (packages)
This step is necessary in order to use external packages. 

**Use the following libraries for the assignment, when needed**:

In [20]:
# ------------>>>>>>>> RUN THIS CODE CELL <<<<<<<<------------
# === CELL TYPE: IMPORTS AND SETUP 

import pandas as pd
import numpy as np

import requests
import json
# ---
import os # for testing only
# ---

# 1. Using a web-API to acquire data
<a id="dataset_desc"></a>
## The data - the quality chocolate database
In this section, You will continue exploring the attributes of quality chocolate.<br />
You will query a database on a distant server, using a web-API.<br/>
This database holds information about quality chocolate in a dedicated table.<br/>

<b>The table consists of the following attributes</b>:
 * company, name, review_date, cocoa_percent, company_location, ,rating, broad_bean_origin, bean_type. 

You can see below a sample of 5 rows:

| company | name | review_date | cocoa_percent | company_location | rating | broad_bean_origin | bean_type |
| :- | :- | :- | :- | :- | :- | :- | :- |
| Castronovo | Tumbes | 2013 | 70 | U.S.A. | 2.75 | Peru | Criollo |
| Blue Bandana | Akesson's E., Sambirano V. | 2016 | 82 | U.S.A. | 3.5 | Madagascar | Trinitario |
| Grenada Chocolate Co. | Grenada | 2008 | 71 | Grenada | 2.5 | Grenada | Trinitario |
| Heirloom Cacao Preservation (Brasstown) | Maya Mtn, Moho R., Toledo D., 2015 | 2016 | 70 | U.S.A. | 3.25 | Belize | Amazon |
| Scharffen Berger | Kumasi Sambirano | 2006 | 68 | U.S.A. | 2 | Ghana & Madagascar | Blend |
| Coppeneur | Los Rios, H. Iara | 2010 | 72 | Germany | 2.5 | Ecuador | Nacional |

## Acquiring the data - about the web-API
In the above [The data - the quality chocolate database](#dataset_desc) section you could see a list of the attributes which you could query, using the web-API.

<b>For each of these attributes we can apply one of the following 2 criteria</b>:
* “<b>refine</b>” – filter rows to match specific parameter values
   * For instance if we use a query which contains the string: <i>refine.rating=3.5</i><br/>
        This means - we want only records, which contain a 'rating' (one of the attribute names) of 3.5 <br/>
        
* “<b>exclude</b>” – exclude rows that contain specific parameter values.
   * For instance if we use a query which contains the string: <i>exclude.company_location=Peru</i><br/>
        This means - we want only records, which DO NOT contain the 'company_location' (one of the attribute names) of Peru <br/>

If we add another constraint, in which we state we want reviews from 2015, then <b>the query string should look as following</b>:
http://3.136.25.169:8080/api/records/1.0/search?dataset=chocolate-bars&exclude.company_location=Peru&refine.review_date=2015&refine.rating=3.5

<b>Web-API URL address notice</b>:
<div class="alert alert-success">
&#x2714; The web-API calls a server located in <a href="http://3.136.25.169:8080">3.136.25.169:8080</a><br />
    &#x2714; We query the chocolate-bars dataset, using: <b>dataset=chocolate-bars</b><br />
&#x2714; The prefix of the query should look like this: <b>http://3.136.25.169:8080/api/records/1.0/search?dataset=chocolate-bars</b><br />
</div>

<b>Note</b>: This web-API uses the rules of the opendatasoft search web-API.<br/>
Please refer to the [Search web-API rules documentation](https://help.opendatasoft.com/apis/ods-search-v1/#refining), for additional information, about the API rules.

## 1. Using a web-API to acquire data

### Instructions
<u>method name</u>: <b>get_res_for_API_query</b>
<pre>The following is expected:
--- Complete the 'get_res_for_API_query' function to get the result from
    the chocolate-bar API, as explained above.
--------
Parameters:
'included_field_value_pairs' - For example, [('review_date', '2015'),
                                     ('company', 'Acalli')] - 
                  means that the 'review_date' must be equal to '2015' and the 'company' 
                  must be equal to 'Acalli'.
                  * Note: the 'included_field_value_pairs' parameter might be empty.
                          In such a case, do not add any 'refine' constraint.      
'excluded_field_value_pairs' - For example, [('company_location', 'Peru')] -
                  means that the 'company_location' MUST NOT be equal to 'Peru'
                  * Note: the 'excluded_field_value_pairs' parameter might be empty.
                          In such a case, do not add any 'exclude' constraint.
--------
The returned value needs to be the object returned from requests.get(str_query)    
</pre>

In [21]:
# 1.
# ------------>>>>>>>> RUN THIS CODE CELL <<<<<<<<------------
# --------  (run after implementation)
# === CODE TYPE: ANSWER 

def get_res_for_API_query(included_field_value_pairs, excluded_field_value_pairs):
    prefix_url = 'http://3.136.25.169:8080/api/records/1.0/search?dataset=chocolate-bars'

    refine_str = '&'.join([f'refine.{field}={val}' for field, val in included_field_value_pairs])
    exclude_str = '&'.join([f'exclude.{field}={val}' for field, val in excluded_field_value_pairs])

    url = '&'.join([prefix_url, refine_str, exclude_str])

    return requests.get(url)

In [22]:
# 1.
# ------------>>>>>>>> RUN THIS CODE CELL <<<<<<<<------------
# --------  (run after implementation)
# === CODE TYPE: SELF TESTING
# Use the following code to test your implementation:

include_field_value_pairs = [('review_date', '2012'), ('company_location', 'Canada')]
exlude_field_value_pairs = [('rating', '3.25')]
res_query = get_res_for_API_query(include_field_value_pairs, exlude_field_value_pairs)
res_json = res_query.json()
res_recs = res_json['records']
res_json

{'nhits': 1234,
 'parameters': {'dataset': 'chocolate-bars',
  'exclude': {'rating': '3.25'},
  'format': 'json',
  'refine': {'company_location': 'Canada', 'review_date': '2012'},
  'rows': 10,
  'timezone': 'UTC'},
 'records': [{'datasetid': 'chocolate-bars',
   'fields': {'bean_type': 'Blend',
    'broad_bean_origin': 'Dom. Rep., Madagascar',
    'cocoa_percent': 70,
    'company': 'Soma',
    'company_location': 'Canada',
    'name': 'Dual Origins, Sambirano, Elvesia',
    'rating': 4.0,
    'ref': '867',
    'review_date': '2012'},
   'record_timestamp': '2016-12-20T18:57:04.946+02:00',
   'recordid': '78a75fece0efe14f0296c2e77737d36a'},
  {'datasetid': 'chocolate-bars',
   'fields': {'bean_type': 'Blend',
    'broad_bean_origin': 'Gre., PNG, Haw., Haiti, Mad',
    'cocoa_percent': 70,
    'company': 'Soma',
    'company_location': 'Canada',
    'name': 'Dancing in Your Head, 5 bean blend',
    'rating': 4.0,
    'ref': '867',
    'review_date': '2012'},
   'record_timestamp': '20

In [21]:
# 1.
# ------------>>>>>>>> RUN THIS CODE CELL <<<<<<<<------------
# --------  (run after implementation, if used)
# === CODE TYPE: SELF TESTING
# Add your additional tests here if needed:

###
### YOUR CODE HERE
###


In [23]:
# 1.
# ------------>>>>>>>> RUN THIS CODE CELL <<<<<<<<------------
# --------  (run only)
# === CODE TYPE: GRADED TEST 

print ("Part 1 - Test 1 (0.5 points) - Sanity (1)")
print ("\t--->Testing the implementation of 'get_res_for_API_query' ...")

include_field_value_pairs = [('review_date', '2012')]
exlude_field_value_pairs = []
res_query = None
try:
    res_query = get_res_for_API_query(include_field_value_pairs, exlude_field_value_pairs)
except Exception as e:
    print ('You probably have a syntax error, we got the following exception:')
    print (str(e))
    raise

print ("Good Job!\nYou've passed the 1st test for the 'get_res_for_API_query' function implementation :-)")

Part 1 - Test 1 (0.5 points) - Sanity (1)
	--->Testing the implementation of 'get_res_for_API_query' ...
Good Job!
You've passed the 1st test for the 'get_res_for_API_query' function implementation :-)


In [24]:
# 1.
# ------------>>>>>>>> RUN THIS CODE CELL <<<<<<<<------------
# --------  (run only)
# === CODE TYPE: GRADED TEST 

print ("Part 1 - Test 2 (0.5 points) - Sanity (2)")
print ("\t--->Testing the implementation of 'get_res_for_API_query' ...")

include_field_value_pairs = []
exlude_field_value_pairs = [('rating', '3.25')]
res_query = None
try:
    res_query = get_res_for_API_query(include_field_value_pairs, exlude_field_value_pairs)
except Exception as e:
    print ('You probably have a syntax error, we got the following exception:')
    print (str(e))
    raise

print ("Good Job!\nYou've passed the 2nd test for the 'get_res_for_API_query' function implementation :-)")

Part 1 - Test 2 (0.5 points) - Sanity (2)
	--->Testing the implementation of 'get_res_for_API_query' ...
Good Job!
You've passed the 2nd test for the 'get_res_for_API_query' function implementation :-)


In [25]:
# 1.
# ------------>>>>>>>> RUN THIS CODE CELL <<<<<<<<------------
# --------  (run only)
# === CODE TYPE: GRADED TEST 

print ("Part 1 - Test 3 (0.5 points) - Sanity (3)")
print ("\t--->Testing the implementation of 'get_res_for_API_query' ...")

include_field_value_pairs = [('review_date', '2012'), ('company_location', 'Canada')]
exlude_field_value_pairs = [('rating', '3.25')]
res_query = None
try:
    res_query = get_res_for_API_query(include_field_value_pairs, exlude_field_value_pairs)
    res_json = res_query.json()
    res_recs = res_json['records']
except Exception as e:
    print ('You probably have a syntax error, we got the following exception:')
    print (str(e))
    raise

print ("Good Job!\nYou've passed the 3rd test for the 'get_res_for_API_query' function implementation :-)")

Part 1 - Test 3 (0.5 points) - Sanity (3)
	--->Testing the implementation of 'get_res_for_API_query' ...
Good Job!
You've passed the 3rd test for the 'get_res_for_API_query' function implementation :-)


In [26]:
# 1.
# ------------>>>>>>>> RUN THIS CODE CELL <<<<<<<<------------
# --------  (run only)
# === CODE TYPE: GRADED TEST 

print ("Part 1 - Test 4 (0.5 points)")
print ("\t--->Testing the implementation of 'get_res_for_API_query' ...")

include_field_value_pairs = [('review_date', '2012'), ('company_location', 'Canada')]
exlude_field_value_pairs = [('rating', '3.25')]
res_query = None
try:
    res_query = get_res_for_API_query(include_field_value_pairs, exlude_field_value_pairs)
    res_json = res_query.json()
    res_recs = res_json['records']
    records_not_dict = [rec for rec in res_recs if type(rec) is not dict]
except Exception as e:
    print ('You probably have a syntax error, we got the following exception:')
    print (str(e))
    raise

assert len(records_not_dict)==0, 'Each of the records should be a dictionary'

print ("Good Job!\nYou've passed the 4th test for the 'get_res_for_API_query' function implementation :-)")

Part 1 - Test 4 (0.5 points)
	--->Testing the implementation of 'get_res_for_API_query' ...
Good Job!
You've passed the 4th test for the 'get_res_for_API_query' function implementation :-)


In [27]:
# 1.
# ------------>>>>>>>> RUN THIS CODE CELL <<<<<<<<------------
# --------  (run only)
# === CODE TYPE: GRADED TEST 

print ("Part 1 - Test 5 (0.5 points)")
print ("\t--->Testing the implementation of 'get_res_for_API_query' ...")

include_field_value_pairs = [('review_date', '2012'), ('company_location', 'Canada')]
exlude_field_value_pairs = [('rating', '3.25')]
res_query = None
try:
    res_query = get_res_for_API_query(include_field_value_pairs, exlude_field_value_pairs)
    res_json = res_query.json()
    res_recs = res_json['records']
    records_not_dict = [rec for rec in res_recs if type(rec) is not dict]
    records_no_fields_key = [rec for rec in res_recs if 'fields' not in rec]
except Exception as e:
    print ('You probably have a syntax error, we got the following exception:')
    print (str(e))
    raise

assert len(records_no_fields_key)==0, "Each of the records should include the key 'fields'"

print ("Good Job!\nYou've passed the 5th test for the 'get_res_for_API_query' function implementation :-)")

Part 1 - Test 5 (0.5 points)
	--->Testing the implementation of 'get_res_for_API_query' ...
Good Job!
You've passed the 5th test for the 'get_res_for_API_query' function implementation :-)


In [28]:
# 1.
# ------------>>>>>>>> RUN THIS CODE CELL <<<<<<<<------------
# --------  (run only)
# === CODE TYPE: GRADED TEST 

print ("Part 1 - Test 6 (0.5 points)")
print ("\t---> - Testing the implementation of 'get_res_for_API_query' ...")

include_field_value_pairs = [('review_date', '2012'), ('company_location', 'Canada')]
exlude_field_value_pairs = [('rating', '3.25')]
res_query = None
try:
    res_query = get_res_for_API_query(include_field_value_pairs, exlude_field_value_pairs)
    res_json = res_query.json()
    res_recs = res_json['records']
    records_not_dict = [rec for rec in res_recs if type(rec) is not dict]
    records_no_fields_key = [rec for rec in res_recs if 'fields' not in rec]
    res_fields = [rec['fields'] for rec in res_recs]
except Exception as e:
    print ('You probably have a syntax error, we got the following exception:')
    print (str(e))
    raise

assert 7==len(res_fields), "Wrong number of result for query"
print ('\n---------------\nFields:')
print (res_fields)

print ("Good Job!\nYou've passed the 6th test for the 'load_dataset' function implementation :-)")

Part 1 - Test 6 (0.5 points)
	---> - Testing the implementation of 'get_res_for_API_query' ...

---------------
Fields:
[{'bean_type': 'Blend', 'broad_bean_origin': 'Dom. Rep., Madagascar', 'cocoa_percent': 70, 'company': 'Soma', 'company_location': 'Canada', 'name': 'Dual Origins, Sambirano, Elvesia', 'rating': 4.0, 'ref': '867', 'review_date': '2012'}, {'bean_type': 'Blend', 'broad_bean_origin': 'Gre., PNG, Haw., Haiti, Mad', 'cocoa_percent': 70, 'company': 'Soma', 'company_location': 'Canada', 'name': 'Dancing in Your Head, 5 bean blend', 'rating': 4.0, 'ref': '867', 'review_date': '2012'}, {'bean_type': 'Trinitario', 'broad_bean_origin': 'Guatemala', 'cocoa_percent': 70, 'company': 'Soma', 'company_location': 'Canada', 'name': 'Cahabon Region', 'rating': 3.5, 'ref': '951', 'review_date': '2012'}, {'bean_type': 'Trinitario', 'broad_bean_origin': 'Papua New Guinea', 'cocoa_percent': 85, 'company': 'Soma', 'company_location': 'Canada', 'name': 'Papua New Guinea', 'rating': 3.5, 'ref':

## 2. Organizing the acquired API information
The records returned from the chocolate opendatasoft API in in [moocdsand](http://moocdsand.ml:8080/api/records/1.0/search) are returned in json format.<br/>
(If the response object is called res_obj) use res_obj.json() to get a dictionary (representing the json).

In this sub section you will organize the results in a dataframe.<br/>
You need to find the relevant path in the dictionary, in order to extract the relevant information.
The relevant information might look as following: 

{"nhits": 2, "parameters": {"dataset": "chocolate-bars", "refine": {"company": "Acalli"}, "timezone": "UTC", "rows": 10, "format": "json"}, "**records**": [{"datasetid": "chocolate-bars", "recordid": "8f51...b42c", "**fields**": {"rating": 3.75, "review_date": "2015", "name": "Tumbes, Norandino", "cocoa_percent": 70, "company": "Acalli", "bean_type": "Criollo", "company_location": "U.S.A.", "broad_bean_origin": "Peru", "ref": "1470"}, "record_timestamp": "2016-12-20T16:57:04.946000+00:00"}, {"datasetid": "chocolate-bars", "recordid": "685b...81d8", "**fields**": {"rating": 3.75, "review_date": "2015", "name": "Chulucanas, El Platanal", "cocoa_percent": 70, "company": "Acalli", "company_location": "U.S.A.", "broad_bean_origin": "Peru", "ref": "1462"}, "record_timestamp": "2016-12-20T16:57:04.946000+00:00"}], "facet_groups": [{"facets": ...

### Instructions
<u>method name</u>: <b>load_query_result_to_df</b>
<pre>The following is expected:
--- Complete the 'load_query_result_to_df' function to load the 'api_query_res' result from the 
    previous 'get_res_for_API_query' to a dataframe.
    Use the api_query_res.json() to get a dictionary (representing the json), as explained above.

    The information of each record (i.e. a row in the dataframe) appears under the 'fields' key,
    in the sub dictionary.
    See the path to 'fields', as showed above.
--------
Parameters:
'included_field_value_pairs' - For example, [('review_date', '2015'),
                                     ('company', 'Acalli')] - 
                  means that the 'review_date' must be equal to '2015' and the 'company' 
                  must be equal to 'Acalli'.
'excluded_field_value_pairs' - For example, [('company_location', 'Peru')] -
                  means that the 'company_location' MUST NOT be equal to 'Peru'
--------
The returned value needs to be the object returned from requests.get(str_query)    
</pre>

In [61]:
# 2.
# ------------>>>>>>>> RUN THIS CODE CELL <<<<<<<<------------
# --------  (run after implementation)
# === CODE TYPE: ANSWER 

def load_query_result_to_df(api_query_res):
    api_query_res_json = api_query_res.json()
    records = api_query_res_json['records']
    fields = [rec['fields'] for rec in records]
    return pd.DataFrame(fields)

In [63]:
# 2.
# ------------>>>>>>>> RUN THIS CODE CELL <<<<<<<<------------
# --------  (run after implementation)
# === CODE TYPE: SELF TESTING
# Use the following code to test your implementation:

include_field_value_pairs = [('review_date', '2013'),
                         ('company_location', 'Belgium')]
exlude_field_value_pairs = [('rating', '3.25'), ('bean_type', 'Trinitario')]
res_query = get_res_for_API_query(include_field_value_pairs, exlude_field_value_pairs)
df_results = load_query_result_to_df(res_query)
df_results


bean_type broad_bean_origin  cocoa_percent        company company_location  \
0   Criollo         Venezuela             74  Benoit Nihant          Belgium   
1   Criollo         Venezuela             74  Benoit Nihant          Belgium   

              name  rating   ref review_date  
0  Cuyagua Village     3.5  1141        2013  
1            Chuao     3.5  1141        2013

In [30]:
# 2.
# ------------>>>>>>>> RUN THIS CODE CELL <<<<<<<<------------
# --------  (run after implementation, if used)
# === CODE TYPE: SELF TESTING
# Add your additional tests here if needed:

###
### YOUR CODE HERE
###


In [64]:
# 2.
# ------------>>>>>>>> RUN THIS CODE CELL <<<<<<<<------------
# --------  (run only)
# === CODE TYPE: GRADED TEST 

print ("Part 2. - Test 1 (0.5 points) - Sanity")
print ("\t--->Testing the implementation of the 'load_query_result_to_df' method (and 'get_res_for_API_query') ...")

include_field_value_pairs = [('review_date', '2013'),
                         ('company_location', 'Belgium')]
exlude_field_value_pairs = [('rating', '3.25'), ('bean_type', 'Trinitario')]

res_query = None
try:
    res_query = get_res_for_API_query(include_field_value_pairs, exlude_field_value_pairs)
    df_results = load_query_result_to_df(res_query)
except Exception as e:
    print ('You probably have a syntax error, we got the following exception:')
    print (str(e))
    raise

print ("Good Job!\nYou've passed the 1st test for the 'load_query_result_to_df' method (and 'get_res_for_API_query') implementation :-)")

Part 2. - Test 1 (0.5 points) - Sanity
	--->Testing the implementation of the 'load_query_result_to_df' method (and 'get_res_for_API_query') ...
Good Job!
You've passed the 1st test for the 'load_query_result_to_df' method (and 'get_res_for_API_query') implementation :-)


In [65]:
# 2.
# ------------>>>>>>>> RUN THIS CODE CELL <<<<<<<<------------
# --------  (run only)
# === CODE TYPE: GRADED TEST 

print ("Part 2. - Test 2 (0.5 points) - Sanity (2)")
print ("\t--->Testing the implementation of the 'load_query_result_to_df' method (and 'get_res_for_API_query') ...")

include_field_value_pairs = [('review_date', '2013'),
                         ('company_location', 'Belgium')]
exlude_field_value_pairs = [('rating', '3.25'), ('bean_type', 'Trinitario')]

res_query = None
try:
    res_query = get_res_for_API_query(include_field_value_pairs, exlude_field_value_pairs)
    df_results = load_query_result_to_df(res_query)
except Exception as e:
    print ('You probably have a syntax error, we got the following exception:')
    print (str(e))
    raise
    
assert type(res_query) is requests.models.Response,'wrong type of result'
assert type(df_results) is pd.DataFrame, 'result should be a dataframe'

print ("Good Job!\nYou've passed the 2nd test for the 'load_query_result_to_df' method (and 'get_res_for_API_query') implementation :-)")

Part 2. - Test 2 (0.5 points) - Sanity (2)
	--->Testing the implementation of the 'load_query_result_to_df' method (and 'get_res_for_API_query') ...
Good Job!
You've passed the 2nd test for the 'load_query_result_to_df' method (and 'get_res_for_API_query') implementation :-)


In [66]:
# 2.
# ------------>>>>>>>> RUN THIS CODE CELL <<<<<<<<------------
# --------  (run only)
# === CODE TYPE: GRADED TEST 

print ("Part 2. - Test 3 (1 points)")
print ("\t--->Testing the implementation of the 'load_query_result_to_df' method (and 'get_res_for_API_query') ...")

include_field_value_pairs = [('review_date', '2013'),
                         ('company_location', 'Belgium')]
exlude_field_value_pairs = [('rating', '3.25'), ('bean_type', 'Trinitario')]


res_query = None
try:
    res_query = get_res_for_API_query(include_field_value_pairs, exlude_field_value_pairs)
    df_results = load_query_result_to_df(res_query)
except Exception as e:
    print ('You probably have a syntax error, we got the following exception:')
    print (str(e))
    raise
    
assert 74==df_results['cocoa_percent'].iloc[0], 'wrong value for cocoa_percent in 1st row'


print ("Good Job!\nYou've passed the 3rd test for the load_query_result_to_df' method (and 'get_res_for_API_query') function implementation :-)")       

Part 2. - Test 3 (1 points)
	--->Testing the implementation of the 'load_query_result_to_df' method (and 'get_res_for_API_query') ...
Good Job!
You've passed the 3rd test for the load_query_result_to_df' method (and 'get_res_for_API_query') function implementation :-)


In [67]:
# 2.
# ------------>>>>>>>> RUN THIS CODE CELL <<<<<<<<------------
# --------  (run only)
# === CODE TYPE: GRADED TEST 

print ("Part 2. - Test 4 (1 points)")
print ("\t--->Testing the implementation of the 'load_query_result_to_df' method (and 'get_res_for_API_query') ...")

include_field_value_pairs = [('review_date', '2013'),
                         ('company_location', 'Belgium')]
exlude_field_value_pairs = [('rating', '3.25'), ('bean_type', 'Trinitario')]


res_query = None
try:
    res_query = get_res_for_API_query(include_field_value_pairs, exlude_field_value_pairs)
    df_results = load_query_result_to_df(res_query)
except Exception as e:
    print ('You probably have a syntax error, we got the following exception:')
    print (str(e))
    raise
    
assert 2==df_results.shape[0], "Wrong number of results for query"
print ('\nResults:')
print (df_results[['name', 'rating', 'company_location', 'cocoa_percent']])

print ("Good Job!\nYou've passed the 4th test for the load_query_result_to_df' method (and 'get_res_for_API_query') function implementation :-)")

Part 2. - Test 4 (1 points)
	--->Testing the implementation of the 'load_query_result_to_df' method (and 'get_res_for_API_query') ...

Results:
              name  rating company_location  cocoa_percent
0  Cuyagua Village     3.5          Belgium             74
1            Chuao     3.5          Belgium             74
Good Job!
You've passed the 4th test for the load_query_result_to_df' method (and 'get_res_for_API_query') function implementation :-)
